src/pip/_internal/operations/prepare.py

```diff
@@ -128,6 +142,7 @@ def unpack_url(
     verbosity: int,
     download_dir: Optional[str] = None,
     hashes: Optional[Hashes] = None,
+    archive_hash: Optional[str] = None,
```

Comment:
Passing this archive_hash to `unpack_url` and then to `get_file_url` is the ugly part.
I think we could clean this up by avoiding calling `unpack_url` when the url is points to a wheel in the wheel cache (as it does not unpack anyway).

If people are ok with general approach I'll investigate that next.

Reply:
Passing `archive_hash` around like this feels wrong, especially `get_file_url` (if it’s passed, a bulk of the logic in the function isn’t used at all). This should be refactored a bit.